### PR TITLE
Dont scroll to revealed files option.

### LINKED ISSUE #23902: Option to keep file selected in sidebar without forcing a scroll
- VSCode Version: 1.10.2
- OS Version: macOS Sierra 10.12.3

This is closely related to #14745.

## Summary: 
I want the selected file in the sidebar to always be my active tab - but **do not** want the sidebar to scroll just because the file is currently not visible in the sidebar.   

That scrolling is the entire reason I turned off `autoReveal` in the first place.

### Options Not Used + Why:  

1. **autoReveal** - (see #14745) automatically scrolling / causing the user to lose their place in the file tree is extremely annoying.
2. **Open Editors Panel** - When there are two files close together in the hierarchy, with the same name, it's too easy to accidentally misread the file paths (especially when they're very long) for that to be a satisfactory replacement for the traditional hierarchy.
3. **Keyboard Shortcut** - [keyboard shortcuts](http://stackoverflow.com/a/42493479) are great, but this should be one of those "don't make me think" things - not a key combo that has to be pressed hundreds of times/day.

### PR DESCRIPTION
**Pull request by phuein**
Adds a new option in settings to not scroll to revealed files in the file explorer view.

This PR fixes #23902

Quick demo: https://drive.google.com/file/d/1k_RL401kYfRgdfC1nVHkyPI65BSL-oXL/view?usp=sharing

### CODE REVIEW

File: src/vs/workbench/contrib/files/browser/files.contribution.ts [hunk 1]
```diff
@@ -375,6 +375,11 @@ configurationRegistry.registerConfiguration({
 			'description': nls.localize('autoReveal', "Controls whether the explorer should automatically reveal and select files when opening them."),
 			'default': true
 		},
+		'explorer.autoRevealNoScroll': {
+			'type': 'boolean',
+			'description': nls.localize('autoRevealNoScroll', "Controls whether the explorer should not automatically scroll to revealed files."),
```
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/files.contribution.ts:380]: Please chagne to "should automatically" the word "not" makes the sentence more confusing

File: src/vs/workbench/contrib/files/browser/files.contribution.ts [hunk 2]
```diff
@@ -375,6 +375,11 @@ configurationRegistry.registerConfiguration({
 			'description': nls.localize('autoReveal', "Controls whether the explorer should automatically reveal and select files when opening them."),
 			'default': true
 		},
+		'explorer.autoRevealNoScroll': {
```
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/files.contribution.ts:378]: `explorer.autoRevealScroll` is a better name

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
@phuein thank you for the PR.
I left some minor comments in your code.
Some things to think about: do users request for other auto reveal behavior and should this setting be a boolean or an open ended string. I suggest you search for all issues labeld with `file-explorer` and search for reveal. And see if there is more things which we could tackle as part of this setting.

However I am not 100% convinced we need this. The deafult behavior imho is fine
* If the file is visible do not scroll when auto revealing
* If explicitly explorer is focused do scroll

What is the case that does not work for you.

**Comment by phuein** [comment]
Hi @isidorn, thanks for reviewing my PR.

I've modified the name and description to fit your suggestions. I agree it's less verbose and confusing this way.

I ran a search for `is:open label:file-explorer reveal` and I'm only seeing a repeat of this request / issue, so it doesn't seem like anything more needs to be added. If more requests come in relating to this issue, I'd be happy to review them, so feel free to tag me.

The case that bothers users is two-fold, and made us disable `explorer.autoReveal` entirely, which loses having the current file highlighted.

1. It may be unwanted for the explorer view to scroll away when closing a file tab. This requires the user to scroll back; repeatedly, when cleaning up tabs.

2. When moving between tabs often, the scrolling / changed project view may become a visual distraction.

**Comment by isidorn** [comment]
Thanks for updating the PR and for claryfing.
However let's go deeper into 1 and 2.

1. So you have a.txt and b.txt open as tabs. Both are visible in the explorer. b.txt is focused, you close it and now a.txt is open. Explorer does not scroll for me. It only scroll if a.txt was not visible. Which is expected. Can you please clarify this?
2. Same as 1. when changing focus from a.txt to b.txt if both are visible no scrolling will happen.

**Comment by clounie** [comment]
@isidorn have you read the original description? It is pretty detailed, this feels like a repeat discussion.

https://github.com/microsoft/vscode/issues/23902#issue-219382529

We just want to see the active file in the sidebar, without the sidebar scrolling on us when our active file changes.

File: src/vs/workbench/contrib/files/browser/files.contribution.ts (hunk 2 above)
Comment (clounie) [reply src/vs/workbench/contrib/files/browser/files.contribution.ts]: What about `explorer.autoScrollOnReveal` ?

As an end user that tells me exactly what's happening without even having to read the description.

File: src/vs/workbench/contrib/files/browser/files.contribution.ts [hunk 3]
```diff
@@ -375,6 +375,11 @@ configurationRegistry.registerConfiguration({
 			'description': nls.localize('autoReveal', "Controls whether the explorer should automatically reveal and select files when opening them."),
 			'default': true
 		},
+		'explorer.autoRevealScroll': {
+			'type': 'boolean',
+			'description': nls.localize('autoRevealScroll', "Controls whether the explorer should automatically scroll to revealed files."),
```
Comment (clounie) [inline src/vs/workbench/contrib/files/browser/files.contribution.ts:380]: Perhaps -
> Controls whether the explorer should automatically scroll when a different file is selected

Comment (phuein) [reply src/vs/workbench/contrib/files/browser/files.contribution.ts:380]: autoRevealScroll depends on autoReveal being active, so I feel the current description fits better.

**Comment by phuein** [comment]
For both 1 and 2, this issue only refers to when the file is not visible in the explorer view, and so causing an unwanted scroll. With this in mind, my previous post explains scenarios where this scrolling may be unwanted.

Video example of scrolling scenario:
https://drive.google.com/file/d/1doXAwszkVNQtj2NUGDw0Fj8kwOkSz_tM/view?usp=sharing

**Comment by isidorn** [comment]
@phuein so why don't you disable autoReveal? Because you want autoReval only when the item is in the viewport?

**Comment by clounie** [comment]
Because if you disable autoReveal, then Explorer does not change the highlighted file.

**Comment by clounie** [comment]
> @phuein so why don't you disable autoReveal? Because you want autoReval only when the item is in the viewport?

Because if you disable autoReveal, then Explorer does not correctly highlight the active file. It just keeps the highlight on the last-highlighted file until the user explicitly clicks on the correct file in Explorer.

**Comment by clounie** [comment]
@isidorn here is a usecase (one of the main usecases) for this:

1. I'm in a Node.js project looking at a folder with 6 different files
2. I Cmd+Click to explore one of my dependencies in node_modules
3. VS Code proceeds to open that node module in explorer, opening a thousand different hierachies in the `node_modules` folder
4. Because of that incredibly annoying autoReveal behavior, I now have to scroll through hundreds if not _thousands_ of folders to close node modules so I can see the rest of my project hierarchy again

*That* is incredibly annoying as a user.

However, I still want to see the active file highlighted (just not scrolled to) because _that_ behavior is very useful.

File: src/vs/workbench/contrib/files/browser/files.contribution.ts (hunk 2 above)
Comment (clounie) [reply src/vs/workbench/contrib/files/browser/files.contribution.ts]: `autoScrollOnReveal` also highlights the dependency on `autoReveal` which you referenced here @phuein https://github.com/microsoft/vscode/pull/96890#discussion_r420491861

**Comment by isidorn** [comment]
Ok thanks for the explanation.
So what does not work for me in this PR is the setting name:
"autoRevealScroll"
If the setting is not scrolled into view it is not revealed, those are synonyms. Also it is funky that the autoRevealScroll depends on another setting autoReveal.

Idealy we would change the `autoReveal` setting to support both boolean values as before and string values.
So I can be `true | false | "on" | "off" | "highlightNoScroll"`.
That way all is under one setting and this feature would be more discoverable.

**Comment by phuein** [comment]
I'm glad the issue and user needs are clarified. I also agree that the previous added setting was funky. I went with it as a simple concept. Modifying autoReveal added complexity. Especially with being boolean backwards compatible. It's now changed and implemented with the new commit.

Thanks for reviewing my work.

**Comment by isidorn** [comment]
Thanks for the changes.
I tried it out and looks good overall.

However there is one problem, now when users who have explorer.autoReveal: true or false will see a warning when they open their settings.json
Unfortunetly we can not ship like this, since thousands of users might see this.
So we need to use a JSON Schema OR type. On the `file.contribution` line 374 change.

Apart from that:
1. You left a `console.log` in the code
2. I do not see the benefit of introducing the `AutoScroll` enum. That way you have to change the code in all the clients. Why not handle it locally in the explorerService so you expect true / false / string. That way changes will be local to that file.

Screenshot 2020-05-11 at 11 28 48

**Comment by phuein** [comment]
Alright, this commit should simplify it. It's either true / false / [string]. Less files and lines are affected. Also fixed the type error for the setting variable. And removed the excess console.log().

**Comment by isidorn** [comment]
Thanks. However looking at the suggested values this is not good. It should suggest true false or the "highlightNoScroll" 
Screenshot 2020-05-12 at 10 22 58

File: src/vs/workbench/contrib/files/browser/files.contribution.ts [hunk 4]
```diff
@@ -371,9 +371,15 @@ configurationRegistry.registerConfiguration({
 			'default': 9
 		},
 		'explorer.autoReveal': {
-			'type': 'boolean',
-			'description': nls.localize('autoReveal', "Controls whether the explorer should automatically reveal and select files when opening them."),
-			'default': true
+			'type': ['boolean', 'string'],
+			'enum': [true, false, 'highlightNoScroll'],
+			'default': 'autoReveal.on',
+			'enumDescriptions': [
```
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/files.contribution.ts:377]: I think this enum descriptions and the default are wrong. Default should be true.

**Review by isidorn — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/files/browser/files.contribution.ts (hunk 4 above)
Comment (phuein) [reply src/vs/workbench/contrib/files/browser/files.contribution.ts:377]: Nice catch, I missed that line. Committed again.

**Comment by isidorn** [comment]
Good job, thanks for this PR. Merging in ☀️ 🍺

**Comment by isidorn** [comment]
Pushed a bit of polish via a06c753e0b

**Comment by phuein** [comment]
Awesome! Thanks for helping out throughout the PR. Everything should be checked now, and I can't wait to enjoy this on release.

**Comment by isidorn** [comment]
Thanks for contributing!

**Pull request merged**